src/plugins/merkleMemberships/server/index.ts

```diff
@@ -17,30 +18,155 @@ import { Router } from 'express';
+
+export type PublicInputArgs = NonEmptyArray<Field>;
+
+const fieldEncDec: EncodeDecoder<FpInterfaceType, Field> = {
+  __interface_tag: 'fp',
+
+  decode: (i: unknown) =>
+    pipe(
+      wrapZodDec('fp', z.string()).decode(i),
+      E.chain((str) => {
+        try {
+          return E.right(Field.from(str));
+        } catch (err) {
+          return E.left('unable to construct field');
+        }
+      })
+    ),
+
+  encode: (i: Field) => i.toString()
+};
+
+const publicInputArgsEncDec: EncodeDecoder<FpInterfaceType, PublicInputArgs> = {
+  __interface_tag: 'fp',
+
+  decode: (i: unknown) =>
+    pipe(
+      wrapZodDec('fp', z.array(z.unknown())).decode(i),
+      E.chain(A.traverse(E.Applicative)(fieldEncDec.decode)),
+      E.chain((arr: Array<Field>) =>
+        E.fromOption(() => 'empty public input')(NE.fromArray(arr))
+      )
+    ),
 
-const PoseidonHashSchema = z.string();
+  encode: NE.map(fieldEncDec.encode)
+};
 
-const publicInputArgsSchema = z.array(PoseidonHashSchema);
+export type Output = {
+  publicInputArgs: PublicInputArgs;
+  recursiveHash: Field;
+};
 
-export type PublicInputArgs = z.infer<typeof publicInputArgsSchema>;
+const outputEncDec: EncodeDecoder<FpInterfaceType, Output> = {
+  __interface_tag: 'fp',
+
+  decode: (i: unknown) =>
+    pipe(
+      E.Do,
+      E.bind('rawObj', () =>
+        wrapZodDec(
+          'fp',
+          z.object({
+            publicInputArgs: z.unknown(),
+            recursiveHash: z.string()
+          })
+        ).decode(i)
+      ),
+      E.bind('publicInputArgs', ({ rawObj }) =>
+        publicInputArgsEncDec.decode(rawObj.publicInputArgs)
+      ),
+      E.bind('recursiveHash', ({ rawObj }) =>
+        fieldEncDec.decode(rawObj.recursiveHash)
+      ),
+      E.map(({ publicInputArgs, recursiveHash }) => {
+        return {
+          publicInputArgs,
+          recursiveHash
+        };
+      })
+    ),
+
+  encode: ({ publicInputArgs, recursiveHash }) => {
+    return {
+      publicInputArgs: publicInputArgsEncDec.encode(publicInputArgs),
+      recursiveHash: fieldEncDec.encode(recursiveHash)
+    };
+  }
+};
+
+type ComputeExpectedHashError =
+  | {
+      __kind: 'tree_missing';
+      root: Field;
+    }
+  | {
+      __kind: 'other';
+      error: string;
+    };
+
+const treeMissingError = (root: Field): ComputeExpectedHashError => {
+  return {
+    __kind: 'tree_missing',
+    root
+  };
+};
+
+const otherError = (error: string): ComputeExpectedHashError => {
+  return {
+    __kind: 'other',
+    error
+  };
+};
+
+const computeExpectedHashErrorToString = (
+  e: ComputeExpectedHashError
+): string =>
+  e.__kind == 'tree_missing'
+    ? `tree with root ${e.root.toString()} not found`
+    : e.error;
+
+const computeExpectedHash =
+  (forest: TreesProvider) =>
+  (roots: NonEmptyArray<Field>): TaskEither<ComputeExpectedHashError, Field> =>
+    pipe(
+      NE.traverse(TE.ApplicativeSeq)((root: Field) =>
+        pipe(
+          forest.getTree(root),
```

Comment:
why getTree returns Option?

Reply:
A tree with the given root might not present in the forest.